# Capstone Project - The Battle of the Neighborhoods (Week 2)
### Applied Data Science Capstone by IBM/Coursera

<a class="anchor" id="introduction"></a>
## Introduction: Business Problem
This project will help solve the problem of where to open a salad shop in the borough of Manhattan in New York City. I will use Foursquare's location data to find out where salad shops are currently located. I will then use a JSON file containing data about the boundaries of neighborhoods to figure out which neighborhoods might have more salad eaters than salad shops. This project would be of interest to the stakeholders of a salad chain such as Chopt, as well as to small business owners who are researching where to open a new location. 

<a class="anchor" id="data"></a>
##Data
The first piece of data that I will need is the latitudinal and longitudinal coordinates of Manhattan, New York. To get this, I will use the module Nominatim from the geopy.geocoders package. I will search Manhattan, New York as an input and will get the latitude and longitude coordinates of Manhattan in return. I will set the returned values to the variables manhattan_latitude and manhattan_longitude, respectively. These variables will be used for the entirety of the project as the central location data. 

I will then be using the Foursquare API to pull data regarding existing salad shops around Manhattan. The Foursquare API will return a dataframe containing the names, addresses, latitudes and longitudes of salad places in the area. 

Finally, I will be using a JSON file containing geospatial data on the boundaries of Manhattan's neighborhoods. I will then use the folium package to visualize which neighborhoods and areas are might benefit from a salad shop.


##Methodology

First, I will use the latitude and longitude of Manhattan, as well as the category ID of salad places to access the Foursquare API. 

Next, I will clean the dataframe from Foursquare to make it easier to work with. I will then use the folium package to plot the locations of each of the salad shops in our dataframe. 

I will then import the JSON file for Manhattan neighborhoods and use it to generate a heat map showing the neighborhoods that have the highest and lowest amount of salad shops. I will then conduct exploratory data analysis to figure out which zip codes in which neighborhoods have the fewest salad shops. 

Import necessary libraries

In [ ]:
import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize

import folium # plotting library

print('Folium installed')
print('Libraries imported.')

Folium installed
Libraries imported.


Set necessary inputs to access the Foursquare API

In [ ]:
CLIENT_ID = 'QEF4SFROUUVOQKNPAFNNRDUW4ACAWSYYG312LM3BKEDDKPIZ' # your Foursquare ID
CLIENT_SECRET = 'CCBFW1WDVTUYDLPFX3SR0QBB5R2UKRYOZ3F1JDJ3PLQMPDRM' # your Foursquare Secret
VERSION = '20180604'
LIMIT = 50
CATEGORYID = '4bf58dd8d48988d1bd941735'
#search_query =
#radius = 

In [ ]:
address = 'Manhattan, NY'
geolocator = Nominatim(user_agent="foursquare_agent")
location = geolocator.geocode(address)
manhattan_latitude = location.latitude
manhattan_longitude = location.longitude
print(manhattan_latitude, manhattan_longitude)

40.7896239 -73.9598939


Call the Foursquare API for a list of salad eateries in Manhattan

In [ ]:
#url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}&categoryId={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, search_query, radius, LIMIT,CATEGORYID)
url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&limit={}&categoryId={}'.format(CLIENT_ID, CLIENT_SECRET, manhattan_latitude, manhattan_longitude, VERSION, LIMIT,CATEGORYID)
url


'https://api.foursquare.com/v2/venues/search?client_id=QEF4SFROUUVOQKNPAFNNRDUW4ACAWSYYG312LM3BKEDDKPIZ&client_secret=CCBFW1WDVTUYDLPFX3SR0QBB5R2UKRYOZ3F1JDJ3PLQMPDRM&ll=40.7896239,-73.9598939&v=20180604&limit=100&categoryId=4bf58dd8d48988d1bd941735'

In [ ]:
results = requests.get(url).json()

In [ ]:
# assign relevant part of JSON to venues
venues = results['response']['venues']

# tranform venues into a dataframe
salad_dataframe = json_normalize(venues)
salad_dataframe.head()


/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:5: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  """


id        name  ... venuePage.id location.neighborhood
0  5961782bf5e9d7031dd38a93  sweetgreen  ...          NaN                   NaN
1  5f3c5b789b9e3e7dcb3b7673  sweetgreen  ...          NaN                   NaN
2  5e1a0a999ab34d00087e4eb6  sweetgreen  ...          NaN                   NaN
3  5d66a35c927f8b0008a59e2b  sweetgreen  ...          NaN                   NaN
4  5b181bcad1a402002cce93ef  Fresh & Co  ...          NaN                   NaN

[5 rows x 25 columns]

In [ ]:
print(salad_dataframe.shape)

(50, 25)


Clean up the dataframe

In [ ]:
salad_joints=salad_dataframe[['name','location.formattedAddress','location.lat','location.lng','location.postalCode']]
salad_joints.columns = ['Name', 'Address', 'Latitude', 'Longitude', 'Zip Code']
salad_joints

Name  ... Zip Code
0               sweetgreen  ...    10025
1               sweetgreen  ...    10019
2               sweetgreen  ...    10003
3               sweetgreen  ...    10016
4               Fresh & Co  ...    10011
5               sweetgreen  ...    10016
6     Bagel Pub Park Slope  ...    11215
7                Bagel Pub  ...    11238
8         Napolini Express  ...    11553
9               sweetgreen  ...    10023
10              sweetgreen  ...    10014
11              Just Salad  ...    10016
12    Xe Máy Sandwich Shop  ...    10009
13              sweetgreen  ...    11201
14              Just Salad  ...    10019
15                   CHOPT  ...    10019
16              sweetgreen  ...    10001
17                    Taïm  ...    10014
18              sweetgreen  ...    11211
19              sweetgreen  ...    10065
20              sweetgreen  ...    11201
21              sweetgreen  ...    10003
22              sweetgreen  ...    10005
23                 Dig Inn  ...    10023
24              Casa Nomad  ...    10001
25                   CHOPT  ...    10012
26              sweetgreen  ...    10003
27              Just Salad  ...    10023
28                   CHOPT  ...    10005
29              sweetgreen  ...    10003
30                    CAVA  ...    10573
31              sweetgreen  ...    10036
32           Pret A Manger  ...    10019
33              sweetgreen  ...    10001
34                   CHOPT  ...    10010
35              sweetgreen  ...    10021
36              sweetgreen  ...    10014
37              sweetgreen  ...    10012
38  Brooklyn Crepe & Juice  ...    11217
39                  Create  ...    11105
40           Pret A Manger  ...    10019
41              Just Salad  ...    10011
42              sweetgreen  ...    10018
43              sweetgreen  ...    10025
44                   CHOPT  ...    10003
45              Just Salad  ...    10010
46              Just Salad  ...    10028
47                   CHOPT  ...    10075
48                   CHOPT  ...    10024
49              sweetgreen  ...    10028

[50 rows x 5 columns]

In [ ]:
salad_joints['Latitude']

0     40.791490
1     40.765232
2     40.726361
3     40.744243
4     40.748508
5     40.746307
6     40.669526
7     40.672343
8     40.722067
9     40.780298
10    40.739183
11    40.749332
12    40.727435
13    40.689910
14    40.762149
15    40.761165
16    40.752706
17    40.735970
18    40.715804
19    40.763714
20    40.702812
21    40.738061
22    40.705626
23    40.774260
24    40.746338
25    40.722076
26    40.733965
27    40.777488
28    40.706239
29    40.729794
30    41.003402
31    40.754640
32    40.765468
33    40.744803
34    40.740738
35    40.767128
36    40.729992
37    40.721184
38    40.679066
39    40.776467
40    40.765453
41    40.741240
42    40.752502
43    40.807284
44    40.737216
45    40.740729
46    40.777037
47    40.773797
48    40.781358
49    40.778012
Name: Latitude, dtype: float64

Map our locations

In [3]:
# create map of New York using latitude and longitude values
map_newyork = folium.Map(location=[manhattan_latitude, manhattan_longitude], zoom_start=10)

# add markers to map
for lat, lng in zip(salad_joints['Latitude'], salad_joints['Longitude']):
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

NameError: ignored

Import the JSON file containing New York's zip code boundaries

In [ ]:
import json # library to handle JSON files
with open('/nyc_zip_code_tabulation_areas_polygons.geojson') as json_data:
    newyork_data = json.load(json_data)
newyork_data

{'features': [{'geometry': {'coordinates': [[[[-73.840759, 40.625364],
       [-73.843062, 40.627933],
       [-73.844219, 40.629957],
       [-73.842498, 40.631778],
       [-73.836267, 40.633592],
       [-73.835921, 40.633354],
       [-73.835506, 40.63264],
       [-73.839544, 40.626779],
       [-73.840688, 40.62723],
       [-73.841587, 40.628103],
       [-73.841813, 40.628011],
       [-73.840845, 40.626662],
       [-73.8401, 40.626436],
       [-73.840759, 40.625364]]]],
    'type': 'MultiPolygon'},
   'properties': {'bldgpostalcode': 0,
    'borough': 'Brooklyn',
    'cartodb_id': 173,
    'cty_fips': '047',
    'id': 'http://nyc.pediacities.com/Resource/PostalCode/11693',
    'objectid': 173,
    'po_name': 'Far Rockaway',
    'postalcode': '11693',
    'shape_area': 3497515.77978,
    'shape_leng': 9479.91727716,
    'st_fips': '36',
    'state': 'NY'},
   'type': 'Feature'},
  {'geometry': {'coordinates': [[[[-73.915441, 40.875591],
       [-73.915087, 40.876496],
       

In [ ]:
data = newyork_data['features']
df = json_normalize(data)
df.head()

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  


type  ...                                      properties.id
0  Feature  ...  http://nyc.pediacities.com/Resource/PostalCode...
1  Feature  ...  http://nyc.pediacities.com/Resource/PostalCode...
2  Feature  ...  http://nyc.pediacities.com/Resource/PostalCode...
3  Feature  ...  http://nyc.pediacities.com/Resource/PostalCode...
4  Feature  ...  http://nyc.pediacities.com/Resource/PostalCode...

[5 rows x 15 columns]

Conduct exploratory data analysis to count how many salad shops are in each zip code

In [ ]:
df2= df[['properties.postalcode']]
df2.columns=['Zip Code']
df2['Contains Salad Place']=0
df2.head(10)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


Zip Code  Contains Salad Place
0    11693                     0
1    10463                     0
2    10033                     0
3    11367                     0
4    10472                     0
5    10037                     0
6    10026                     0
7    10119                     0
8    10044                     0
9    10036                     0

In [ ]:
all_zip_codes=df['properties.postalcode'].tolist()
zip_with_salad=salad_joints['Zip Code'].tolist()
print("There are {} zip codes total!".format(str(len(all_zip_codes))))

There are 262 zip codes total!


In [ ]:
from collections import Counter
Count=Counter(zip_with_salad)


In [ ]:
for x in range(0,len(all_zip_codes)):
   if df2['Zip Code'][x] in zip_with_salad:
     df2['Contains Salad Place'][x]=Count[df2['Zip Code'][x]]



/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2882: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  exec(code_obj, self.user_global_ns, self.user_ns)


In [ ]:
df2.head(50)

Zip Code  Contains Salad Place
0     11693                     0
1     10463                     0
2     10033                     0
3     11367                     0
4     10472                     0
5     10037                     0
6     10026                     0
7     10119                     0
8     10044                     0
9     10036                     1
10    10002                     0
11    10038                     0
12    11206                     0
13    10275                     0
14    10111                     0
15    10022                     0
16    10043                     0
17    10170                     0
18    10112                     0
19    10122                     0
20    10107                     0
21    10103                     0
22    10081                     0
23    10304                     0
24    10169                     0
25    10167                     0
26    10177                     0
27    10172                     0
28    10310                     0
29    11212                     0
30    10302                     0
31    11096                     0
32    11416                     0
33    10470                     0
34    10469                     0
35    10039                     0
36    10029                     0
37    10010                     2
38    11411                     0
39    11040                     0
40    10096                     0
41    11220                     0
42    11225                     0
43    11209                     0
44    11228                     0
45    11372                     0
46    11004                     0
47    10305                     0
48    11096                     0
49    11223                     0

In [ ]:
print(max(df2['Contains Salad Place']))

5


Create a heat map with our data

In [ ]:
world_geo = r'/content/nyc_zip_code_tabulation_areas_polygons.geojson' # geojson file

    # create a plain world map
world_map = folium.Map(location=[40.693943, -73.985880], zoom_start=12, tiles='Mapbox Bright')

# generate choropleth map using the total immigration of each country to Canada from 1980 to 2013
folium.Choropleth(
    geo_data=world_geo,
    data=df2,
    columns=['Zip Code', 'Contains Salad Place'],
    key_on='feature.properties.postalcode',
    fill_color='YlOrRd', 
    fill_opacity=0.7, 
    line_opacity=0.2,
    legend_name='Number of Salad Places Per Neighborhood'
).add_to(world_map)

# display map

world_map

Find which zip codes have the fewest salad shops

In [ ]:
sorted(Count.items(), key=lambda kv: kv[1], reverse=False)

[('10573', 1),
 ('11211', 1),
 ('10065', 1),
 ('11105', 1),
 ('10009', 1),
 ('10021', 1),
 ('11217', 1),
 ('11553', 1),
 ('10024', 1),
 ('11238', 1),
 ('11215', 1),
 ('10075', 1),
 ('10018', 1),
 ('10036', 1),
 ('10025', 2),
 ('10005', 2),
 ('10011', 2),
 ('10012', 2),
 ('11201', 2),
 ('10010', 2),
 ('10028', 2),
 ('10016', 3),
 ('10023', 3),
 ('10001', 3),
 ('10014', 3),
 ('10003', 5),
 ('10019', 5)]

##Results
Our results showed that the zip codes 10573, 11211, 10065, 11105, 10009, 10021, 11217, 11553, 10024, 11238, 11215, 10075, 10018, 10036 all only contain a single salad shop. 

Our results also showed that the zip codes 10003, 10019 in Gramercy Park and West Midtown both have 5 salad shops each. These are the most of any zip code we looked at. 

##Obervations
Our results showed that the densest concentrations of salad shops are below central park in midtown Manhattan. The areas with the sparsest concentration of salad shops are lower Manhattan, around the financial districts and the lower east and west sides. 





##Conclusion

In conclusion, I would advise looking further into opening a salad shop in lower Manhattan. This area has the sparsest amount of salad shops. It is also a highly populated area of the city where there are a lot of office buildings where people may be interested in getting a salad for lunch. 